### ISSUE TITLE
Trainer.evaluate and Trainer.metrics_names do not interact as specified in the docs

### ISSUE THREAD

**Issue by RyanSaxe** [keras-team/keras#21487]
From the docstring of evaluate's return in the base trainer class:

>         Returns:
>             Scalar test loss (if the model has a single output and no metrics)
>             or list of scalars (if the model has multiple outputs
>             and/or metrics). The attribute `model.metrics_names` will give you
>             the display labels for the scalar outputs.

However, the implementation is inconsistent with `model.metric_names`

```python
@property
def metrics_names(self):
    return [m.name for m in self.metrics]
```

The above implementation DOES NOT get the submetrics for compiled metrics. What this means is that, if you have multiple compiled metrics, `model.evaluate` and `model.metrics_names` will not be lists of the same size, which is what is expected from the docstring shared above.

Looking at the base Trainer class, I can see the function `_flatten_metrics_in_order` has more of what I would expect. And looking at the tensorflow Trainer evaluate, this function is exactly what gets called for the output of evaluate (which is why it does not match the docs):

```python
def _flatten_metrics_in_order(self, logs):
    """Turns `logs` dict into a list as per key order of `metrics_names`."""
    metric_names = []
    for metric in self.metrics:
        if isinstance(metric, CompileMetrics):
            metric_names += [
                sub_metric.name for sub_metric in metric.metrics
            ]
        else:
            metric_names.append(metric.name)
    # the rest of the implementation here is irrelevant
```

Hence, to interact correctly with `evaluate`, the below should be the implementation of `metrics_names`

```python
@property
def metrics_names(self):
    metric_names = []
    for metric in self.metrics:
        if isinstance(metric, CompileMetrics):
            metric_names += [
                sub_metric.name for sub_metric in metric.metrics
            ]
        else:
            metric_names.append(metric.name)
    return metric_names
```

However, since `metrics_names` is a part of the exposed API, this could easily break other peoples code if they depend on it. So maybe this should be exposed through a different route?

Regardless, I'll make my own utility function for now for my needs, but seems that this should be addressed directly in keras eventually. Maybe even just with a docs update?

**Comment by sonali-kumari1** [comment]
Hi @RyanSaxe - 
Thanks for reporting this issue. It appears similar to issue [#20244](https://github.com/keras-team/keras/issues/20244). 
I have tested the reported behavior with a small example in this [gist](https://colab.sandbox.google.com/gist/sonali-kumari1/9d8548135711679d9d7730d53f24b2f4/-21487.ipynb) and here is the ouput:
```
model.metrics_names: ['loss', 'compile_metrics']
model.evaluate() returned: [0.08282306045293808, 0.24390719830989838, 0.08282306045293808]
```
This confirms that `model.metrics_names `only shows top-level names like ['loss', 'compile_metrics'], while `evaluate()` returns all metric values including submetrics, causing a size mismatch. We will look into this update you!

**Comment by monicadsong** [comment]
Hi @RyanSaxe -- 

The Keras team would like to learn more your Keras usage. Please thumbs up this comment if you are open to speaking with us and we'll reach out to you over LinkedIn. 

Thanks!
Monica, Keras Product Manager
